### ISSUE TITLE
Error parsing Cargo.toml file due to ZoD schema validation error

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#32528]
### Discussed in https://github.com/renovatebot/renovate/discussions/32523

<div type='discussions-op-text'>

<sup>Originally posted by **rami3l** November 14, 2024</sup>
### How are you running Renovate?

A Mend.io-hosted app

### Please tell us more about your question or problem

Coming from Rustup, we are using the following configuration WRT `cargo` dependencies:

```json
        {
            "matchManagers": [
                "cargo"
            ],
            "matchUpdateTypes": [
                "patch"
            ],
            "enabled": false
        },
        {
            "matchManagers": [
                "cargo"
            ],
            "matchUpdateTypes": [
                "minor"
            ],
            "matchCurrentVersion": "!/^0/",
            "enabled": false
        },
```
_https://github.com/rust-lang/rustup/blob/c8eddcb4c1dc9aea8067fa73179ce1791aa24d1d/.github/renovate.json#L24-L41_

... and we set the lock file maintenance to be executed every Monday. It **used to be working alright**, at least till the point of 2024/10/28 (https://github.com/rust-lang/rustup/pull/4066), but after that I can no longer see `cargo` on the dependency dashboard (https://github.com/rust-lang/rustup/issues/3142, you can refer to the editing history of that post as well). The logs say the same:

```json
INFO: Dependency extraction complete
{
  "baseBranch": "master"
  "stats": {
    "managers": {
      "dockerfile": {
        "fileCount": 23,
        "depCount": 23
      },
      "github-actions": {
        "fileCount": 10,
        "depCount": 107
      }
    },
    "total": {
      "fileCount": 33,
      "depCount": 130
    }
  }
}
```

Many thanks in advance!
</div>

**Comment by rarkins** [comment]
I added the rustup toml/lock to fixtures in https://github.com/renovatebot/renovate/pull/32529 so that we have a "failing" case to then fix

**Comment by rami3l** [comment]
Looking at:

> It seems to be caused by a Zod validation error:
> 
> ```console
> DEBUG: Error parsing Cargo.toml file (repository=rust-lang/rustup, packageFile=Cargo.toml)
> ...
>          "issues": {"package": {"version": "Expected string, received object"}}
> ```
> _https://github.com/renovatebot/renovate/discussions/32523#discussioncomment-11250115_

The direct cause of this issue on our side seems to be the migration to workspace-wide properties (https://github.com/rust-lang/rustup/pull/4061/commits/72e2e767a2ceb0d5c905700546abaf0d11edd7bb).

**Comment by viceice** [comment]
- #31260
- #32533

**Comment by viceice** [comment]
> > Sorry for that, I've attempted a fix in #32530. This basically comes from:
> > ```toml
> > [package]
> > version.workspace = true
> > ```
> >   
> > The schema always expects a string for `package.version`, did not notice that it could also be an object.
> 
> @mkniewallner I'm not sure about how this project consumes manifest schemata in general, but we do have a schema for `Cargo.toml` right now, if it helps: https://github.com/rust-lang/cargo/blob/3ad481e749f64e43041d4c7145906b5ef1c16532/crates/cargo-util-schemas/manifest.schema.json

Nice, @zharinov So you know a way to validate a zod schema agains a json schema? that would probably helpful

**Comment by rami3l** [comment]
@viceice Thanks a lot for the hard work! May I ask when I am expected to get this fix on Mend.io-hosted instances though?
